sequence_transform.c

```diff
@@ -181,23 +182,65 @@ void st_handle_repeat_key()
     }
 }
 //////////////////////////////////////////////////////////////////////////////////////////
+void log_rule(st_trie_t *trie, st_trie_payload_t *res) {
+    char context_string[SEQUENCE_MAX_LENGTH];
+    st_key_buffer_get_context_string(&key_buffer, context_string, res->context_len);
+
+    char rule_trigger_char = context_string[res->context_len - 1]; 
+    context_string[res->context_len - 1] = '\0';
+
+    bool is_repeat = rule_trigger_char == 'R' && strlen(context_string) == 0;
+
+    if (is_repeat) {
+        context_string[0] = st_keycode_to_char(st_key_buffer_get_keycode(&key_buffer, -2));
+        context_string[1] = '\0';
+    }
+
+    uprintf("st_rule,%s,%d,%c,", context_string, res->num_backspaces, rule_trigger_char);
+
+    const uint16_t completion_end = res->completion_index + res->completion_len;
+
+    for (uint16_t i = res->completion_index; i < completion_end; ++i) {
+        char ascii_code = CDATA(i);
+        uprintf("%c", ascii_code);
+    }
+
+    switch (res->func_code) {
+        case 1:  // repeat
+            if (is_repeat)
+                uprintf("%s", context_string);
+            break;
+
+        case 2:  // set one-shot shift
+            uprintf("%c", 'S');
+            break;
+    }
+
+    uprintf("\n");
+}
+//////////////////////////////////////////////////////////////////////////////////////////
 void st_handle_result(st_trie_t *trie, st_trie_payload_t *res) {
 #ifdef SEQUENCE_TRANSFORM_LOG_GENERAL
     uprintf("completion search res: index: %d, len: %d, bspaces: %d, func: %d\n",
             res->completion_index, res->completion_len, res->num_backspaces, res->func_code);
 #endif
+
 #if defined(RECORD_RULE_USAGE) && defined(CONSOLE_ENABLE)
-    uprintf("used rule: hello world\n");
+    log_rule(trie, res);
 #endif
+
```

Comment:
Please avoid style changes to code that is unrelated to the new features. (Here: these new empty lines.)